src/v/cluster/tm_stm.cc

```diff
@@ -697,6 +697,8 @@ ss::future<tm_stm::op_status> tm_stm::do_register_new_producer(
         co_return tm_stm::op_status::unknown;
     }
 
+    _cache->set_mem(tx.etag, tx_id, tx);
```

Comment:
If i understand the tx cache part correctly this should not be required. And it would only cover up the source of the problem. Here we call `replicate_quorum_ack` right before executing this, hence updating an in memory state is not required as producer id is already persisted in the log and should be available on the new leader after it is fully up to date

Reply:
The problem, as I understand it, arises from the fact that we don't record changes to the transaction in between it is created and committed. We just bump etag and put it to cache. This is explained in more detail in #7591.

Now, when the leader changes, it does indeed first lookup it in the log and finds it (based on the transaction id only) https://github.com/redpanda-data/redpanda/blob/f1b96f0ca19d4c932417a5214874f158d7b7d022/src/v/cluster/tx_gateway_frontend.cc#L2570

But then, it needs to know the "true" state of the transaction which might have been changed in between it was written to the log and `tx_gateway_frontend::get_tx` was called (e.g. as part of the AddPartitionsToTxn client request). Remember, we only recorded the fact that transaction was created, all subsequent changes aren't recorded durably anywhere.

This is where we run a fetch_tx round and try to find an in-memory tx state (which is stored on the leader at the time of last tx change) based on the (txid, etag) https://github.com/redpanda-data/redpanda/blob/f1b96f0ca19d4c932417a5214874f158d7b7d022/src/v/cluster/tx_gateway_frontend.cc#L2645

Reply:
I’m not convinced that this change is necessary, and it seems risky as it violates one of the invariants of the tx subsystem.

**Why do I think it’s unnecessary?**

The function `do_register_new_producer` replicates an update to the log using `replicate_quorum_ack`. It then waits for the update to be applied to the state machine with `wait_no_throw`, which only returns control flow after `tm_stm::apply` is invoked. You can verify that the latter method updates the cache via `_cache->set_log` from `apply_tm_update`.

Therefore, when the additional statement `_cache->set_mem` is executed, the update should already be in the cache.

**Which invariant is being violated?**

The invocation of `set_mem` is not expected after `set_log` for the same update. Moreover, when `set_log` is called, it explicitly erases the in-memory cache record.

Reply:
it looks like `set_log` only erases the record when its etag is older than the current one, in this case it will not be removed. 
Maybe then there is a bug in `tm_stm_cache::find` method as currently when entry is not found in `_state` it returns an empty optional.
[](https://github.com/redpanda-data/redpanda/blob/2fa7227ef35c2436d16387d664c7946fb7aad376/src/v/cluster/tm_stm_cache.cc#L50C3-L50C3)

`_state` is only populated when calling `set_mem` so this is the only way the `tx_id` can be in cache when it is present in `_state` map.

Reply:
The way I'm reading the code is that we have 2 caches. One is "mem", the in-memory cache. Another, is "log", a lookup table for the latest state that was persisted in the log. The "log" is only updated for initial state and the final state of the transaction. All the intermediary states are only stored in the "mem".

`tx_gateway_frontend::get_tx` (called as part of all tx RPCs) does few things to find the latest state of a transaction, the important ones for this conversation are the following steps:
1. It tries to fetch the latest state from "mem", if not found then from "log".
2. If tx term is less than the current "stm" term, then it means that no interactions were made with the transaction on this node and we need to find where the last interaction did happen.
   
   In the first paragraph, I called out _"All the intermediary states are only stored in the "mem"_. As a result, we MUST find an in-memory state of the transaction. If it can't be found we can't make any assumptions. We can't make an assumption that (not found == initial state) as the transaction could made progress but the node restarted and lost the cache.
  
    As a consequence of this requirement, we SHALL at some point have a "set_mem" call. The remaining question is where do we put it.

    Option A: As part of `set_log` vs. Option B: Somewhere after `set_log`.

    Option A doesn't work for the following reasons:
        
      1. `set_log` is called when doing log replay/snapshot apply. Imagine a transaction that made progress, then the node was restarted and during replay we would update the "mem" with tx state == "initial". This would effectively make the transaction travel back in time.
      2. `set_log` is called on all replicas a part of apply. If all replicas call `set_mem`, then we interact with the tx on the leader, (optionally leader failing), future fetch_tx calls will find the ready state tx on one of the replicas. Effectively, we travel back in time again. Having a leader fail is optional to trigger a problem. fetch_tx seems to succeed on the first successful reply. So we can travel back in time just by transferring leadership and old leader being slow to reply. [`tx_seq` is compared to avoid this being a problem](https://github.com/redpanda-data/redpanda/blob/b8f63a76bc1f9943503a61951321381a6f62013a/src/v/cluster/tx_gateway_frontend.cc#L2872), but we might end up aborting/forgetting transaction based on luck and timings, not nice.

    This leaves us with Option B.

    > The invocation of set_mem is not expected after set_log for the same update. Moreover, when set_log is called, it explicitly erases the in-memory cache record.

    Why is this invariant important? What can go wrong? To me it seems that `set_mem` for the same `set_log` update is required for the system to function correctly.